File: static/js/ecdh.js
const p = BigInt('0xffffffff00000001000000000000000000000000ffffffffffffffffffffffff');
const a = BigInt('0xffffffff00000001000000000000000000000000fffffffffffffffffffffffc');
const b = BigInt('0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b');
const Gx = BigInt('0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296');
const Gy = BigInt('0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5');
const n = BigInt('0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551');

/** 
 * Returns a Promise resolving to a Uint8Array of the SHA-256 hash of the input key.
 * @param {Uint8Array} key 
 * @returns {Promise<Uint8Array>}
 */
function sha256(key) {
	return crypto.subtle.digest('SHA-256', key).then(hashBuffer => new Uint8Array(hashBuffer));
}

/**
 * Modulu function so that the answer won't be negative (n % m).
 * @param {number} n 
 * @param {number} m 
 * @returns 
 */
function mod(n, m) {
	return ((n % m) + m) % m;
}

/** 
 * Modular inverse using Extended Euclidean Algorithm.
 * @param {number} a
 * @param {number} m
 */
function modinv(a, m) {
	a = mod(a, m);
	let m0 = m;
	let x0 = 0n, x1 = 1n;

	if (m === 1n) return 0n;

	while (a > 1n) {
		const q = a / m;
		[a, m] = [m, a % m];
		[x0, x1] = [x1 - q * x0, x0];
	}

	if (x1 < 0n) x1 += m0;
	return x1;
}

/**
 * Converts a non-negative BigInt to a Uint8Array of specified length in big-endian order.
 * Big-endian means the most significant byte is at the lowest index (start) of the array.
 * @param {*} n - The bigint.
 * @param {*} byteLength - the length of the output byte array.
 * @returns {Uint8Array} A byte array representing the bigint in big-endian byte order.
 */
function bigintToBytes(n, byteLength = 32) {
	if (n < 0n) throw new Error("Only non-negative BigInts supported");
	const bytes = new Uint8Array(byteLength);
	for (let i = byteLength - 1; i >= 0 && n > 0n; i--) {
		bytes[i] = Number(n & 0xFFn);
		n >>= 8n;
	}
	return bytes;
}

export class Point {
	constructor(x, y, curve) {
		this.x = x;
		this.y = y;
		this.curve = curve;

		if (!this.isAtInfinity() && !curve.isOnCurve(this)) {
			throw new Error(`Point (${x}, ${y}) is not on the given curve`);
		}
	}

	isAtInfinity() {
		return this.x === null && this.y === null;
	}

	/** Checks if two points are the same. */
	equals(other) {
		return (
			this.x === other.x &&
			this.y === other.y &&
			this.curve === other.curve
		);
	}

	/** 
	 * Adds two points on the elliptic curve.
	 * @param {Point} other - The point to add.
	 * @returns {Point} The resulting point.
	 */
	add(other) {
		const { a, p } = this.curve;

		if (this.curve !== other.curve) {
			throw new Error("Cannot add points on different curves");
		}

		if (this.isAtInfinity()) return other;
		if (other.isAtInfinity()) return this;

		if (this.x === other.x && (this.y !== other.y || this.y === 0n)) {
			return new Point(null, null, this.curve); // Point at infinity
		}

		// The slope:
		let m;

		if (this.equals(other)) {
		    // Point doubling (שיפוע משיק)
			m = mod((3n * this.x ** 2n + a) * modinv(2n * this.y, p), p);
		} else {
			// Point addition (שיפוע הישר העובר בשתי הנקודות)
			m = mod((other.y - this.y) * modinv(other.x - this.x, p), p);
		}

		const xR = mod(m * m - this.x - other.x, p);
		const yR = mod(m * (this.x - xR) - this.y, p);

		return new Point(xR, yR, this.curve);
	}


	/** Computes k * P using double-and-add algorithm.
	 * @param {BigInt} k - The scalar multiplier.
	 * @returns {Point} The resulting point.
	 * @throws {Error} If k is not in the range [1, n-1].
	*/
	scalarMult(k) {
		if (k < 1n || k >= this.curve.n) {
			throw new Error("k must be in the range [1, n-1]");
		}

		let result = new Point(null, null, this.curve);

		// Start with the current point as the addend
		let addend = this;

		// Loop through each bit of k (from least significant to most significant)
		while (k > 0n) {
			// If the current least significant bit is 1, add the addend to the result
			if (k & 1n) result = result.add(addend);
			
			// Double the point for the next bit
			addend = addend.add(addend);

			// Shift k right by 1 bit
			k >>= 1n;
		}

		return result;
	}

	toString() {
		if (this.isAtInfinity()) {
			return "Point(infinity)";
		}
		return `Point(${this.x}, ${this.y})`;
	}
}


class ECDH {
	constructor(a, b, p, n, Gx, Gy) {
		this.a = a;
		this.b = b;
		this.p = p;
		this.n = n;
		this.G = new Point(Gx, Gy, this);

		if (!this.isOnCurve(this.G)) {
			throw new Error("Base point G is not on the curve");
		}

		if (this.G.scalarMult(n - 1n).isAtInfinity() === true) {
			throw new Error("(n-1)*G is infinity — invalid n");
		}
	}

	/** Checks if a point is on the curve. */
	isOnCurve(point) {
		if (point.isAtInfinity()) return true;
		const { x, y } = point;
		const left = mod(y ** 2n, this.p);
		const right = mod(x ** 3n + this.a * x + this.b, this.p);
		return left === right;
	}

	/** 
	 * Generates a random private key in the range [1, n-1].
	 * @returns {BigInt} The private key.
	*/
	generatePrivateKey() {
		const byteLength = (this.n.toString(2).length + 7) >> 3;

		while (true) {
			const rand = new Uint8Array(byteLength);
			crypto.getRandomValues(rand);

			let num = 0n;
			for (const byte of rand) {
				num = (num << 8n) + BigInt(byte);
			}

			if (num > 0n && num < this.n) return num;
		}
	}

	/** Generates the public key from the private key.
	 * @param {BigInt} privateKey - The private key.
	 * @returns {Point} The public key.
	 * @throws {Error} If the private key is not in the range [1, n-1].
	*/
	generatePublicKey(privateKey) {
		return this.G.scalarMult(privateKey);
	}

	/** Generates a shared key using the private key and the other party's public key.
	 * @param {BigInt} privateKey - The private key.
	 * @param {Point} otherPublicKey - The other party's public key.
	 * @returns {Promise<Uint8Array>} The shared key (32 bytes).
	 * @throws {Error} If the private key is not in the range [1, n-1].
	 */
	generateSharedKey(privateKey, otherPublicKey) {
		const sharedSecret = otherPublicKey.scalarMult(privateKey);

		if (sharedSecret.isAtInfinity()) {
			throw new Error("Shared secret is at infinity");
		}

		// Convert x and y to 32-byte big-endian arrays
		const xBytes = bigintToBytes(sharedSecret.x, 32);
		const yBytes = bigintToBytes(sharedSecret.y, 32);

		// Concatenate x and y bytes
		const sharedSecretBytes = new Uint8Array([...xBytes, ...yBytes]);

		// Return the sha256 hash (32 bytes Uint8Array)
		return sha256(sharedSecretBytes);
	}

	/** 
	 * Generates a nonce with the first 12 bytes of the key.
	 * @param {Uint8Array} sharedKey - The shared key (32 bytes).
	 * @returns {Uint8Array} The nonce (12 bytes).
	*/
	generateSharedNonce(sharedKey) {
		if (sharedKey.length !== 32) {
			throw new Error("Shared key must be 32 bytes");
		}
		return sharedKey.subarray(0, 12);
	}

	toString() {
		return `ECDH(a=${this.a}, b=${this.b}, p=${this.p}, n=${this.n}, G=(${this.G.x}, ${this.G.y}))`;
	}
}

export const ecdh = new ECDH(a, b, p, n, Gx, Gy);

File: server_crypto/ecdh.py
# !!! If you run _example.py do this and not line 3:
# from secp256r1 import *
from server_crypto.secp256r1 import *
import hashlib
import secrets

# This code implements the elliptic curve Diffie Hellman (ECDH) key exchange algorithm using the secp256r1 curve.
# It allows two computers to securely share a secret key over the network.

class Point:
    """Represents a point on the elliptic curve."""

    def __init__(self, x, y, curve):
        self._x = x
        self._y = y
        self._curve = curve

        if not self.is_at_infinity() and not curve.is_on_curve(self):
            raise ValueError(f"Point ({x}, {y}) is not on the given curve")

    @property
    def x(self):
        return self._x
    
    @property
    def y(self):
        """Y coordinate of the point."""
        return self._y
    
    @property
    def curve(self):
        return self._curve
    
    def is_at_infinity(self):
        return self.x is None and self.y is None

    def __eq__(self, other):
        """Checks if two points are the same."""
        return (self.x, self.y, self.curve) == (other.x, other.y, other.curve)

    def __add__(self, other):
        """Adds two points on the elliptic curve."""
        if self.curve != other.curve:
            raise ValueError("Cannot add points on different curves")

        if self.is_at_infinity():
            return other
        if other.is_at_infinity():
            return self
        if self.x == other.x and (self.y != other.y or self.y == 0):
            return Point(None, None, self.curve)  # point at infinity

        p = self.curve.p

        # m is the slope
        if self == other:
            # point doubling
            m = (3 * self.x ** 2 + self.curve.a) * pow(2 * self.y, -1, p)
        else:
            # point addition
            m = (other.y - self.y) * pow(other.x - self.x, -1, p)

        m %= p
        x_r = (m * m - self.x - other.x) % p
        y_r = (m * (self.x - x_r) - self.y) % p
        return Point(x_r, y_r, self.curve)

    def scalar_mult(self, k: int):
        """Computes k * P using double-and-add algorithm."""
        if k < 1 or k >= self.curve.n:
            raise ValueError("k must be in the range [1, n-1]")
        result = Point(None, None, self.curve)

        # Start with the current point as the addend
        addend = self

        # Loop through each bit of k (from least significant to most significant)
        while k:
            # If the current least significant bit is 1, add the addend to the result
            if k & 1:
                result += addend

            # Double the point for the next bit
            addend += addend

            # Shift k right by 1 bit
            k >>= 1

        return result
    
    def __repr__(self):
        """String representation of the point."""
        if self.is_at_infinity():
            return f"Point(infinity)"
        return f"Point({self.x}, {self.y})"
    

class ECDH:
    """Elliptic Curve Diffie Hellman key exchange."""

    def __init__(self, a: int, b: int, p: int, n: int, Gx: int, Gy: int):
        self.a = a
        self.b = b
        self.p = p
        self.n = n
        self.G = Point(Gx, Gy, self)

        if not self.is_on_curve(self.G):
            raise ValueError("Base point G is not on the curve")

        if self.G.scalar_mult(self.n - 1).is_at_infinity():
            raise ValueError("(n-1)*G is infinity — invalid n")

    def is_on_curve(self, point: Point):
        """Check if a point is on the curve by placing the point in 
        the equasion and checking if it is true."""
        if point.is_at_infinity():
            return True
        x, y = point.x, point.y
        return (y * y - x ** 3 - self.a * x - self.b) % self.p == 0
    
    def generate_private_key(self):
        """Generate a secure random private key with the secrets library."""
        return secrets.randbelow(self.n - 1) + 1

    def generate_public_key(self, private_key: int):
        """Generate the public key by multiplying the private key with the base point G."""
        if private_key < 1 or private_key >= self.n:
            raise ValueError("Private key must be in the range [1, n-1]")
        
        return self.G.scalar_mult(private_key)

    @staticmethod
    def generate_shared_key(private_key: int, other_public_key: Point) -> bytes:
        """
        Generates the shared secret key by multiplying the private key with the 
        other's public key, and then hashing x, y using sha256 to get a large key.
        :param private_key: (int) The private key. 
        :param other_public_key: (Point) The public key of the other person.
        :return: (bytes) A 32-byte shared secret key
        """
        shared_secret = other_public_key.scalar_mult(private_key)
        
        if shared_secret.is_at_infinity():
            raise ValueError("Shared secret is at infinity")
        
        # Combine the bits of x and y coordinates of the shared secret point
        shared_secret_bytes = (
            shared_secret.x.to_bytes(32, byteorder='big') + 
            shared_secret.y.to_bytes(32, byteorder='big')
        )
        
        # Returns the hashed shared secret bytes
        return hashlib.sha256(shared_secret_bytes).digest()
    
    @staticmethod
    def generate_shared_nonce(shared_key: bytes) -> bytes:
        """
        Generates the nonce for the chacha20 cipher by taking the first 
        12 bytes of the shared key.
        :return: (bytes) A 12-byte nonce.
        """
        return shared_key[:12]
    
    def __repr__(self):
        """String representation of the ECDH object."""
        return f"ECDH(a={self.a}, b={self.b}, p={self.p}, n={self.n}, G=({self.G.x}, {self.G.y}))"


ecdh = ECDH(a, b, p, n, Gx, Gy)

if __name__ == "__main__":
    # Client and Server generate their private keys
    client_private_key = ecdh.generate_private_key()
    server_private_key = ecdh.generate_private_key()

    # Client and Server generate their public keys
    client_public_key = ecdh.generate_public_key(client_private_key)
    server_public_key = ecdh.generate_public_key(server_private_key)

    # Client computes the shared secret using Server's public key and its private key
    shared_secret_client = ecdh.generate_shared_key(client_private_key, server_public_key)

    # Server computes the shared secret using Client's public key and its private key
    shared_secret_server = ecdh.generate_shared_key(server_private_key, client_public_key)

    # Verify that both shared secrets are the same
    assert shared_secret_client == shared_secret_server, "Shared keys do not match!"

    # Print the 256-bit key (in int format)
    print(f"Shared key (bytes): {shared_secret_client}")
    print(f"Shared base nonce (bytes): {ecdh.generate_shared_nonce(shared_secret_client)}")
